finance-reports/rev_exp_report_to_s3.py

```diff
@@ -0,0 +1,306 @@
+Runs the current month's report and last month's report.
+"""
+import argparse
+import calendar
+from datetime import datetime
+from io import StringIO
+import os
+import logging
+
+import boto3
+from mstrio.api.reports import report_instance, get_prompted_instance
+from mstrio.project_objects.report import Report
+from mstrio.connection import Connection
+
+import utils
+
+BASE_URL = os.getenv("BASE_URL")
+MSTRO_USERNAME = os.getenv("MSTRO_USERNAME")
+MSTRO_PASSWORD = os.getenv("MSTRO_PASSWORD")
+PROJECT_ID = os.getenv("PROJECT_ID")
+
+# AWS Credentials
+AWS_ACCESS_ID = os.getenv("AWS_ACCESS_KEY")
+AWS_SECRET_ACCESS_KEY = os.getenv("AWS_SECRET_ACCESS_KEY")
+BUCKET = os.getenv("BUCKET_NAME")
+
+EXP_REPORT_ID = "1C804F8891479811944EF68F99835649"
+REV_REPORT_ID = "FBC5E5F30744717D7079ADADB956C3BC"
+
+conn = Connection(
+    base_url=BASE_URL,
+    username=MSTRO_USERNAME,
+    password=MSTRO_PASSWORD,
+    project_id=PROJECT_ID,
+    login_mode=1,
+)
+
+
+def select_month(year, month):
+    """
+    Parameters
+    ----------
+    year : Int
+        Argument provided value for year.
+    month : Int
+        Argument provided value for month.
+    Returns
+    -------
+    f_year : int
+        Selected Year
+    f_month : int
+        Selected Month
+    """
+    # If args are missing, default to current month and/or year
+    if not year:
+        f_year = datetime.now().year
+    else:
+        f_year = year
+
+    if not month:
+        f_month = datetime.now().month
+    else:
+        f_month = month
+
+    return f_year, f_month
+
+
+def get_month_date(year, month):
+    """
+    Returns the last day of the month as a string for the given year-month integers
+    """
+    # Get last day of the month
+    last_day = calendar.monthrange(year, month)[1]
+    # Convert to string
+    return datetime(year, month, last_day).strftime("%Y-%m-%d")
+
+
+def get_fiscal_year(year, month):
+    if month >= 10:
+        fiscal_year = year + 1
+    else:
+        fiscal_year = year
+
+    return fiscal_year
+
+
+def build_todos(year, month, date_str):
+    """
+    Returns a list of dictionaries of required parameters to run microstrategy reports
+    """
+    if month == 1:
+        prev_month = 12
+        prev_year = year - 1
+    else:
+        prev_month = month - 1
+        prev_year = year
+
+    prev_date = get_month_date(prev_year, prev_month)
+    # ATD = 2400, TPW = 6200
+    departments = ["2400", "6200"]
+    months = [prev_month, month]
+    fys = [get_fiscal_year(prev_year, prev_month), get_fiscal_year(year, month)]
+    dates = [prev_date, date_str]
+
+    todos = []
+    item = {}
+    for dept in departments:
+        for i in range(0, 2):
+            item["department"] = dept
+            item["month"] = months[i]
+            item["fy"] = fys[i]
+            item["date"] = dates[i]
+            todos.append(item)
+            item = {}
+
+    return todos
+
+
+def expenses_prompts(fy, date, dept, prompts):
```

Comment:
cool - thanks for making it clear in the code that prompts are a a Microstrategy thing. 🙏 i looked into [them](https://www2.microstrategy.com/producthelp/Current/ReportDesigner/WebHelp/Lang_1033/Content/prompts.htm) a little because I was curious.